IPNS.md

```diff
@@ -29,95 +36,258 @@ Each time a file is modified, its content address changes. As a consequence, the
+and refer to IPNS addresses as `/ipns/{ipns-name}` (or `/ipns/{libp2p-key}`).
+
+## IPNS Record
+
+A logical IPNS record is a data structure containing the following fields:
+
+- **Value** (bytes)
+  - It can be any path, such as a `/ipns/{ipns-key}` path to another IPNS record, a [DNSLink](https://dnslink.dev/) path (`/ipns/example.com`) or an immutable IPFS path (`/ipfs/baf...`).
+  - Implementations MUST include this value in both `IpnsEntry.value` and inside the DAG-CBOR document in `IpnsEntry.data[value]`.
+- **Validity Type** (uint64)
+  - Defines the conditions under which the record is valid.
+  - The only supported value is `0`, which indicates the `validity` field contains the expiration date after which the IPNS record becomes invalid.
+  - Implementations MUST support `validityType = 0` and include this value in both `IpnsEntry.validityType` and inside the DAG-CBOR document at `IpnsEntry.data[validityType]`.
+- **Validity** (bytes)
+  - When `validityType = 0`
+    - Expiration date of the record with nanoseconds precision.
+    - Represented as an ASCII string that follows notation from [RFC3339](https://www.ietf.org/rfc/rfc3339.txt) (`1970-01-01T00:00:00.000000001Z`).
+  - Implementations MUST include this value in both `IpnsEntry.validity` and inside the DAG-CBOR document at `IpnsEntry.data[validity]`.
+- **Sequence** (uint64)
+  - Represents the current version of the record (starts at 0).
+  - Implementations MUST include this value in both `IpnsEntry.sequence` and inside the DAG-CBOR document at `IpnsEntry.data[sequence]`.
+- **TTL** (uint64)
+  - A hint for how long the record should be cached before going back to, for instance the DHT, in order to check if it has been updated.
+  - Implementations MUST include this value in both `IpnsEntry.ttl` and inside the DAG-CBOR document at `IpnsEntry.data[ttl]`.
+- **Public Key** (bytes)
+  - Public key used to sign this record.
+    - If public key is small enough to fit in IPNS name (e.g., Ed25519 keys inlined using `identity` multihash), `IpnsEntry.pubKey` field is redundant and MAY be skipped to save space.
+    - The public key MUST be included if it cannot be extracted from the IPNS name (e.g., legacy RSA keys). Implementers MUST follow key serialization defined in [PeerID specs](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#key-types).
+- **Signature** (bytes)
+  - Provides the cryptographic proof that the IPNS record was created by the owner of the private key.
+  - Implementations MUST include this value in `IpnsEntry.signatureV2` and follow signature creation and verification as described in [Record Creation](#record-creation) and [Record Verification](#record-verification).
+- **Extensible Data** (DAG-CBOR)
+  - Extensible record data in [DAG-CBOR](https://ipld.io/specs/codecs/dag-cbor/spec/) format.
+  - The default set of fields can be augmented with additional information. Implementations are free to leverage this, or simply ignore unexpected fields.
```

Comment:
Probably obvious, but might be worth flagging that any new fields that aren't added to the spec could run into collisions and/or misinterpretations (e.g. like custom HTTP headers).

Also, technically you can add more data to the protobuf as well unless we're speccing that implementations should be stricter in how they interpret the protobufs (e.g. no adding field number 100 into the protobuf) it's just not covered by the signature so not something most users would want.

Allowing for new protobuf fields does give us some extensibility if we want to modify the format more into the future though, so probably worth keeping that around so long as there are record size limits in place.

Reply:
Added lower size limit https://github.com/ipfs/specs/pull/319/commits/cc5941b55ee0af7f05f83cb84e9451dd786f3842, think this is easier from implementation perspective (does not require custom protobuf implementation, and the size check of entire thing is very inexpensive).

Reply:
@lidel Maybe the spec should recommend applications make their own field with sub-fields under it, e.g. `IpnsEntry.data.my-app.field`?

Reply:
Thanks, added best practices in https://github.com/ipfs/specs/pull/319/commits/78c9f045cb79ed7587ec8ed889dc2d90e4b486f4